packages/core/src/fields/types/file/index.ts

```diff
@@ -1,128 +1,117 @@
-    },
-  }),
-});
-
-const FileFieldOutput = graphql.interface<FileData>()({
+const FileFieldOutput = graphql.object<FileMetadata & { storage: string }>()({
   name: 'FileFieldOutput',
-  fields: fileFields,
-  resolveType: val => (val.mode === 'local' ? 'LocalFileFieldOutput' : 'CloudFileFieldOutput'),
-});
-
-const LocalFileFieldOutput = graphql.object<FileData>()({
-  name: 'LocalFileFieldOutput',
-  interfaces: [FileFieldOutput],
-  fields: fileFields,
-});
-
-const CloudFileFieldOutput = graphql.object<FileData>()({
-  name: 'CloudFileFieldOutput',
-  interfaces: [FileFieldOutput],
-  fields: fileFields,
+  fields: {
+    filename: graphql.field({ type: graphql.nonNull(graphql.String) }),
+    filesize: graphql.field({ type: graphql.nonNull(graphql.Int) }),
+    url: graphql.field({
+      type: graphql.nonNull(graphql.String),
+      resolve(data, args, context) {
+        return context.files(data.storage).getUrl(data.filename);
+      },
+    }),
+  },
 });
 
-async function inputResolver(data: FileFieldInputType, context: KeystoneContext) {
+async function inputResolver(storage: string, data: FileFieldInputType, context: KeystoneContext) {
   if (data === null || data === undefined) {
-    return { mode: data, filename: data, filesize: data };
-  }
-
-  if (data.ref) {
-    if (data.upload) {
-      throw userInputError('Only one of ref and upload can be passed to FileFieldInput');
-    }
-    return context.files!.getDataFromRef(data.ref);
-  }
-  if (!data.upload) {
-    throw userInputError('Either ref or upload must be passed to FileFieldInput');
+    return { filename: data, filesize: data };
   }
   const upload = await data.upload;
-  return context.files!.getDataFromStream(upload.createReadStream(), upload.filename);
+  return context.files(storage).getDataFromStream(upload.createReadStream(), upload.filename);
 }
 
 export const file =
   <ListTypeInfo extends BaseListTypeInfo>(
-    config: FileFieldConfig<ListTypeInfo> = {}
+    config: FileFieldConfig<ListTypeInfo>
   ): FieldTypeFunc<ListTypeInfo> =>
-  () => {
-    if ((config as any).isIndexed === 'unique') {
+  meta => {
+    const storage = meta.assets.getStorage(config.storage);
+
+    if (!storage) {
+      throw new Error(
+        `${meta.listKey}.${meta.fieldKey} has storage set to ${config.storage} but there is no storage config under that key`
+      );
+    }
+
+    if ('isIndexed' in config) {
       throw Error("isIndexed: 'unique' is not a supported option for field type file");
     }
 
     return fieldType({
       kind: 'multi',
       fields: {
         filesize: { kind: 'scalar', scalar: 'Int', mode: 'optional' },
-        mode: { kind: 'scalar', scalar: 'String', mode: 'optional' },
         filename: { kind: 'scalar', scalar: 'String', mode: 'optional' },
       },
     })({
       ...config,
+      hooks: storage.preserve
+        ? config.hooks
+        : {
+            ...config.hooks,
+            async beforeOperation(args) {
```

Comment:
@dcousens you talked about doing this as `afterOperations` in some code channel spitballing - was there a conclusion there?